src/crimson/admin/admin_socket.cc

```diff
@@ -0,0 +1,693 @@
+
+  Neither gate closure, nor a failure of the code block we are running, will
+  cause an exception to be thrown by safe_action_gate_func()
+*/
+template <typename AsyncAction>
+seastar::future<> do_until_gate(seastar::gate& gt, AsyncAction action)
+{
+  auto stop_cond = [&gt]() { return gt.is_closed(); };
+  auto safe_action{ [as1 = std::move(action), &gt]() mutable {
+    if (gt.is_closed())
+      return seastar::make_ready_future<>();
+    return with_gate(gt, [as1 = std::move(as1)]() {
+      return as1().handle_exception([](auto e) {});
+    });
+  } };
+
+  return seastar::do_until(stop_cond, std::move(safe_action)).discard_result();
+}
+
+seastar::future<> AdminSocket::init(const std::string& path)
+{
+  logger().debug("{}: asok socket path={}", __func__, path);
+  auto sock_path = seastar::socket_address{ seastar::unix_domain_addr{ path } };
+
+  std::ignore = internal_hooks().discard_result().then([this, sock_path] {
+    return seastar::do_with(
+      seastar::engine().listen(sock_path),
+      [this](seastar::server_socket& lstn) {
+        m_server_sock = &lstn;  // used for 'abort_accept()'
+        return do_until_gate(arrivals_gate,
+                             [&lstn, this]() {
+                               return lstn.accept().then(
+                                 [this](seastar::accept_result from_accept) {
+                                   seastar::connected_socket cn =
+                                     std::move(from_accept.connection);
+
+                                   return do_with(
+                                     std::move(cn.input()),
+                                     std::move(cn.output()), std::move(cn),
+                                     [this](auto& inp, auto& out, auto& cn) {
+                                       return handle_client(inp, out).finally(
+                                         [this, &inp, &out] {
+                                           ;  // left for debugging: std::cerr
+                                              // << "finally "  << std::endl;
+                                         });
+                                     });
+                                 });
+                             })
+          .then([this] {
+            logger().debug("{}: admin-sock thread terminated", __func__);
+            return seastar::now();
+          });
+      });
+  });
+
+  return seastar::make_ready_future<>();
+}
+
+seastar::future<> AdminSocket::stop()
+{
+  if (m_server_sock && !arrivals_gate.is_closed()) {
+    // note that we check 'is_closed()' as if already closed - the server-sock
+    // may have already been discarded
+    m_server_sock->abort_accept();
+    m_server_sock = nullptr;
+  }
+  return seastar::futurize_apply([this] { return arrivals_gate.close(); })
+    .then_wrapped([this](seastar::future<> res) {
+      if (res.failed()) {
+        std::ignore = res.handle_exception([](std::exception_ptr eptr) {
```

Comment:
I don't think this is necessary.  I think you can just let res fall out of scope?

Reply:
AFAIK I will get a log warning from Seastar whenever I remove an exceptional future. I actually added such "just ignore it all" code following such log warnings.